cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,875 @@
+  withAffine affine $ \affinePtr ->
+    withNewP' $ \pp ->
+      c_blst_from_affine pp affinePtr
+
+-- | Infinity check on curve points.
+isInf :: (BLS_Curve curve) => P curve -> Bool
+isInf p = unsafePerformIO $ withP p c_blst_p_is_inf
+
+affineInG :: (BLS_Curve curve) => Affine curve -> Bool
+affineInG affine = unsafePerformIO $
+  withAffine affine c_blst_affine_in_g
+
+generator :: (BLS_Curve curve) => P curve
+generator = unsafePFromPPtr c_blst_generator
+
+---- Scalar / Fr operations
+
+scalarFromFr :: Fr -> IO Scalar
+scalarFromFr fr =
+  withNewScalar' $ \scalarPtr ->
+    withFr fr $ \frPtr ->
+      c_blst_scalar_from_fr scalarPtr frPtr
+
+frFromScalar :: Scalar -> IO Fr
+frFromScalar scalar =
+    withNewFr' $ \frPtr ->
+      withScalar scalar $ \scalarPtr ->
+        c_blst_fr_from_scalar frPtr scalarPtr
+
+frFromCanonicalScalar :: Scalar -> IO (Maybe Fr)
+frFromCanonicalScalar scalar
+  | scalarCanonical scalar
+  = Just <$> frFromScalar scalar
+  | otherwise
+  = return Nothing
+
+scalarFromBS :: ByteString -> Either BLSTError Scalar
+scalarFromBS bs = unsafePerformIO $ do
+  BS.useAsCStringLen bs $ \(cstr, l) ->
+    if l == sizeScalar then do
+      (success, scalar) <- withNewScalar $ \scalarPtr ->
+        c_blst_scalar_from_be_bytes scalarPtr cstr (fromIntegral l)
+      if success then
+        return $ Right scalar
+      else
+        return $ Left BLST_BAD_SCALAR
+    else
+      return $ Left BLST_BAD_SCALAR
+
+scalarToBS :: Scalar -> ByteString
+scalarToBS scalar = unsafePerformIO $ do
+  cstr <- mallocForeignPtrBytes sizeScalar
+  withForeignPtr cstr $ \cstrp -> do
+    withScalar scalar $ \scalarPtr -> do
+      c_blst_bendian_from_scalar cstrp scalarPtr
+      BS.packCStringLen (castPtr cstrp, sizeScalar)
+
+scalarCanonical :: Scalar -> Bool
+scalarCanonical scalar = unsafePerformIO $
+  withScalar scalar c_blst_scalar_fr_check
+
+---- PT operations
+
+ptMult :: PT -> PT -> PT
+ptMult a b = unsafePerformIO $
+  withPT a $ \ap ->
+    withPT b $ \bp ->
+      withNewPT' $ \cp ->
+        c_blst_fp12_mul cp ap bp
+
+ptEq :: PT -> PT -> Bool
+ptEq a b = unsafePerformIO $
+  withPT a $ \ap ->
+    withPT b $ \bp ->
+      c_blst_fp12_is_equal ap bp
+
+ptFinalVerify :: PT -> PT -> Bool
+ptFinalVerify a b = unsafePerformIO $
+  withPT a $ \ap ->
+    withPT b $ \bp ->
+      c_blst_fp12_finalverify ap bp
+
+instance Eq PT where
+  (==) = ptEq
+
+---- Pairings
+
+pairing :: P1 -> P2 -> Either BLSTError PT
+pairing p1 p2 =
+  if inGroup p1 && inGroup p2 then
+    Right . unsafePerformIO $
+      withAffine (toAffine p1) $ \ap1 ->
+        withAffine (toAffine p2) $ \ap2 ->
+          withNewPT' $ \ppt ->
+            c_blst_miller_loop ppt ap2 ap1
+  else
+    Left BLST_POINT_NOT_IN_GROUP
+
+---- Utility
+
+withMaybeCStringLen :: Maybe ByteString -> (CStringLen -> IO a) -> IO a
+withMaybeCStringLen Nothing go = go (nullPtr, 0)
+withMaybeCStringLen (Just bs) go = BS.useAsCStringLen bs go
```

Comment:
It would be useful to add some documentation stating that the supplied action must not mutate the pointer. This is already true because of `nullPtr` usage
```suggestion
withMaybeCStringLen (Just bs) go = BS.unsafeUseAsCStringLen bs go
```